plugin/pkg/auth/authorizer/node/graph.go

```diff
@@ -0,0 +1,244 @@
+type Graph struct {
+	lock  sync.RWMutex
+	graph *simple.DirectedAcyclicGraph
+	nodes map[nodeType]map[string]map[string]*NamedNode
+}
+
+func NewGraph() *Graph {
+	return &Graph{
+		nodes: map[nodeType]map[string]map[string]*NamedNode{},
+		graph: simple.NewDirectedAcyclicGraph(0, 0),
+	}
+}
+
+type nodeType byte
+
+const (
+	ConfigMapNode nodeType = iota
+	NodeNode
+	PodNode
+	PVCNode
+	PVNode
+	SecretNode
+)
+
+var nodeTypes = map[nodeType]string{
+	ConfigMapNode: "configmap",
+	NodeNode:      "node",
+	PodNode:       "pod",
+	PVCNode:       "pvc",
+	PVNode:        "pv",
+	SecretNode:    "secret",
+}
+
+// must be called under a write lock
+func (g *Graph) getOrCreateNode(nodeType nodeType, namespace, name string) *NamedNode {
+	if node, exists := g.getNode(nodeType, namespace, name); exists {
+		return node
+	}
+	return g.createNode(nodeType, namespace, name)
+}
+
+// must be called under a read lock
+func (g *Graph) getNode(nodeType nodeType, namespace, name string) (*NamedNode, bool) {
+	node, exists := g.nodes[nodeType][namespace][name]
+	return node, exists
+}
+
+// must be called under a write lock
+func (g *Graph) createNode(nodeType nodeType, namespace, name string) *NamedNode {
+	typedNodes, exists := g.nodes[nodeType]
+	if !exists {
+		typedNodes = map[string]map[string]*NamedNode{}
+		g.nodes[nodeType] = typedNodes
+	}
+
+	namespacedNodes, exists := typedNodes[namespace]
+	if !exists {
+		namespacedNodes = map[string]*NamedNode{}
+		typedNodes[namespace] = namespacedNodes
+	}
+
+	node := NewNode(nodeType, namespace, name, g.graph.NewNodeID())
+	namespacedNodes[name] = node
+	g.graph.AddNode(node)
+
+	return node
+}
+
+// must be called under write lock
+func (g *Graph) deleteNode(nodeType nodeType, namespace, name string) {
+	node, exists := g.getNode(nodeType, namespace, name)
+	if !exists {
+		return
+	}
+
+	// find existing neighbors
+	neighborsToRemove := []graph.Node{}
+	g.graph.VisitFrom(node, func(neighbor graph.Node) bool {
+		if g.graph.Degree(neighbor) == 1 {
+			neighborsToRemove = append(neighborsToRemove, neighbor)
+		}
+		return true
+	})
+	g.graph.VisitTo(node, func(neighbor graph.Node) bool {
+		if g.graph.Degree(neighbor) == 1 {
+			neighborsToRemove = append(neighborsToRemove, neighbor)
+		}
+		return true
+	})
+
+	// remove the node
+	g.graph.RemoveNode(node)
+	delete(g.nodes[nodeType][namespace], name)
+	if len(g.nodes[nodeType][namespace]) == 0 {
+		delete(g.nodes[nodeType], namespace)
+	}
+	if len(g.nodes[nodeType]) == 0 {
+		delete(g.nodes, nodeType)
+	}
+
+	// remove neighbors that are now edgeless
+	for _, neighbor := range neighborsToRemove {
+		g.graph.RemoveNode(neighbor)
+		n := neighbor.(*NamedNode)
+		delete(g.nodes[n.nodeType][n.namespace], n.name)
+		if len(g.nodes[n.nodeType][n.namespace]) == 0 {
+			delete(g.nodes[n.nodeType], n.namespace)
+		}
+		if len(g.nodes[n.nodeType]) == 0 {
+			delete(g.nodes, n.nodeType)
+		}
+	}
```

Comment:
no, we're only watching pods and pvs to populate this graph, and inferring the existence of configmaps/secrets, pvcs, and nodes by the references from the pods and pvs. if I remove a pod, I want to remove any secrets it referenced if nothing else is referencing them.